httpx/_transports/asgi.py

```diff
@@ -120,38 +191,76 @@ async def receive() -> dict:
             return {"type": "http.request", "body": body, "more_body": True}
 
         async def send(message: dict) -> None:
-            nonlocal status_code, response_headers, body_parts
-            nonlocal response_started, response_complete
+            nonlocal status_code, response_headers
 
             if message["type"] == "http.response.start":
-                assert not response_started
+                assert not response_started.is_set()
 
                 status_code = message["status"]
                 response_headers = message.get("headers", [])
-                response_started = True
+                response_started.set()
 
             elif message["type"] == "http.response.body":
                 assert not response_complete.is_set()
                 body = message.get("body", b"")
                 more_body = message.get("more_body", False)
 
                 if body and method != b"HEAD":
-                    body_parts.append(body)
+                    await produce_response_body_chunk(body)
 
                 if not more_body:
                     response_complete.set()
 
-        try:
-            await self.app(scope, receive, send)
-        except Exception:
-            if self.raise_app_exceptions or not response_complete:
-                raise
+        async def run_app() -> None:
+            nonlocal app_exception
+            try:
+                await self.app(scope, receive, send)
+            except Exception as exc:
+                app_exception = exc
+                app_crashed.set()
+
+        aclose_app = await create_background_task(run_app)
+
+        await run_until_first_complete(app_crashed.wait, response_started.wait)
 
-        assert response_complete.is_set()
+        if app_crashed.is_set():
+            assert app_exception is not None
+            await aclose_app(app_exception)
+            if self.raise_app_exceptions or not response_started.is_set():
+                raise app_exception
+
+        assert response_started.is_set()
         assert status_code is not None
         assert response_headers is not None
 
-        stream = ByteStream(b"".join(body_parts))
+        async def aiter_response_body_chunks() -> typing.AsyncIterator[bytes]:
+            chunk = b""
+
+            async def consume_chunk() -> None:
+                nonlocal chunk
+                chunk = await consume_response_body_chunk()
+
+            while True:
+                await run_until_first_complete(
+                    app_crashed.wait, consume_chunk, response_complete.wait
+                )
```

Comment:
Just a curious note - Having played with Golang a bit recently, it's interesting to see how this pattern of "run some functions concurrently until one is done, and then figure out what happened" sort of matches using channels and `select { case ... }` in Go. (Which makes me think it's a generally useful and correct pattern.)